File: apps/ios/ChatApp/Mesh/MeshScale.swift
import Foundation

/// MeshScale.swift — mesh scaling: coverage grows with device count.
///
/// Mirrors services/mesh/scale.go. The packet TTL (max hops) is the primary
/// lever. A larger network needs a larger hop budget to span its diameter.

enum MeshScale {
    static let defaultMaxHops = 64

    static func hopsForDevices(_ n: Int) -> Int {
        guard n > 0 else { return defaultMaxHops }
        var hops = Int(ceil(6.0 * sqrt(Double(n))))
        if hops < 8 { hops = 8 }
        if hops > 1024 { hops = 1024 }
        return hops
    }
}

File: apps/android/app/src/main/java/com/chatapp/mesh/MeshScale.kt
package com.chatapp.mesh

import kotlin.math.*
// MeshScale.kt — scaling helpers for the offline mesh.
//
// Mirrors services/mesh/scale.go. The packet TTL (max hops) is the primary
// lever. A larger network needs a larger hop budget to span its logical
// diameter.

object MeshScale {
    const val DEFAULT_MAX_HOPS = 64

    /** Returns a hop budget sized for a mesh of [n] devices, driven by sqrt. */
    fun hopsForDevices(n: Int): Int {
        if (n <= 0) return DEFAULT_MAX_HOPS
        var hops = ceil(6.0 * sqrt(n.toDouble())).toInt()
        if (hops < 8) hops = 8
        if (hops > 1024) hops = 1024
        return hops
    }
}